cedar-lean/Cedar/Thm/Data/List.lean

```diff
@@ -841,73 +915,323 @@ theorem forall₂_implies_all_right {α β} {R : α → β → Prop} {xs : List
+      simp only [h₃, Except.bind_err, Except.bind_ok] at h₁
+      rename_i ytl
+      simp only [Except.ok.injEq] at h₁
+      subst h₁
+      exact List.Forall₂.cons h₂ (ih h₃)
+  case mpr =>
+    intro h₁
+    induction xs generalizing ys
+    case nil =>
+      simp only [forall₂_nil_left_iff] at h₁
+      simp only [mapM'_nil, pure, Except.pure, h₁]
+    case cons xhd xtl ih =>
+      simp only [mapM'_cons, pure, Except.pure]
+      replace ⟨yhd, ytl, h₁, h₃, h₄⟩ := forall₂_cons_left_iff.mp h₁
+      subst ys
+      cases h₂ : f xhd
+      case error e => simp [h₁] at h₂
+      case ok y' =>
+        simp only [h₁, Except.ok.injEq] at h₂
+        subst y'
+        specialize @ih ytl h₃
+        simp only [ih, Except.bind_err, Except.bind_ok]
+
+theorem mapM_ok_iff_forall₂ {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
+  List.mapM f xs = .ok ys ↔
+  List.Forall₂ (λ x y => f x = .ok y) xs ys
+:= by
+  rw [← List.mapM'_eq_mapM]
+  exact mapM'_ok_iff_forall₂
+
+/--
+  Note that the converse is not true:
+  counterexample `xs` is `[1]`, `ys` is `[1, 2]`, `f` is `Except.ok`
 
+  But for a limited converse, see `all_ok_implies_mapM'_ok`
+-/
 theorem mapM'_ok_implies_all_ok {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
   List.mapM' f xs = .ok ys →
   ∀ x ∈ xs, ∃ y ∈ ys, f x = .ok y
 := by
   intro h
-  exact forall₂_implies_all_left (mapM'_ok_implies_forall₂ h)
+  exact forall₂_implies_all_left (mapM'_ok_iff_forall₂.mp h)
+
+theorem mapM_ok_implies_all_ok {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
+  List.mapM f xs = .ok ys →
+  ∀ x ∈ xs, ∃ y ∈ ys, f x = .ok y
+:= by
+  rw [← List.mapM'_eq_mapM]
+  exact mapM'_ok_implies_all_ok
 
+theorem all_ok_implies_mapM'_ok {α β γ} {f : α → Except γ β} {xs : List α} :
+  (∀ x ∈ xs, ∃ y, f x = .ok y) →
+  ∃ ys, List.mapM' f xs = .ok ys
+:= by
+  intro h₁
+  induction xs
+  case nil => exists []
+  case cons xhd xtl ih =>
+    simp only [mem_cons, forall_eq_or_imp] at h₁
+    replace ⟨⟨yhd, h₁⟩, h₂⟩ := h₁
+    replace ⟨ytl, ih⟩ := ih h₂
+    exists yhd :: ytl
+    simp [h₁, ih, pure, Except.pure]
+
+theorem all_ok_implies_mapM_ok {α β γ} {f : α → Except γ β} {xs : List α} :
+  (∀ x ∈ xs, ∃ y, f x = .ok y) →
+  ∃ ys, List.mapM f xs = .ok ys
+:= by
+  rw [← List.mapM'_eq_mapM]
+  exact all_ok_implies_mapM'_ok
+
+/--
+  Note that the converse is not true:
+  counterexample `ys` is `[1]`, `xs` is `[1, 2]`, `f` is `Except.ok`
+
+  But for a limited converse, see `all_from_ok_implies_mapM'_ok`
+-/
 theorem mapM'_ok_implies_all_from_ok {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
   List.mapM' f xs = .ok ys →
   ∀ y ∈ ys, ∃ x ∈ xs, f x = .ok y
 := by
   intro h
-  exact forall₂_implies_all_right (mapM'_ok_implies_forall₂ h)
+  exact forall₂_implies_all_right (mapM'_ok_iff_forall₂.mp h)
 
-theorem mapM'_some_implies_forall₂ {α β} {f : α → Option β} {xs : List α} {ys : List β} :
-  List.mapM' f xs = .some ys →
-  List.Forall₂ (λ x y => f x = .some y) xs ys
```

Comment:
I needed the converse as well, so I made this an iff and renamed it `mapM'_some_iff_forall2`. This lemma was introduced in https://github.com/cedar-policy/cedar-spec/pull/297. I reformulated the proofs in https://github.com/cedar-policy/cedar-spec/pull/291 (and here) to rely on this lemma instead of the other lemmas in https://github.com/cedar-policy/cedar-spec/pull/291.

Reply:
If we're removing or renaming existing lemmas, let's leave the old name and signature in place, marked as `@[deprecated]`.  See [here](https://github.com/leanprover/std4/blob/f58165d3d6e0b048d89e56944e98d9054b223d9b/Std/Data/BitVec/Lemmas.lean#L13-L14) for an example.

This will make it easier to fix the internal code.  Without that, Lean tells us that the old name is missing, but it can be hard to figure out / remember what's supposed to replace the old name in a 1,000+ LOC file :) 

Once the internal code is patched, I can put up a new PR that removes all the deprecated lemmas.

Alternatively, if we don't want to deal with `@[deprecated]` tags, let's adopt some commenting convention that's easy to grep for / remove, which can help figure out the mapping of old names to new names.

Reply:
Very reasonable.  Did this for this theorem and the one I commented on above; let me know if there are other instances where we need a deprecated alias.